doc/lua_api.txt

```diff
@@ -6911,16 +6924,65 @@ of the `${k}` syntax in formspecs is not deprecated.
 * `set_float(key, value)`
 * `get_float(key)`: Returns `0` if key not present.
 * `get_keys()`: returns a list of all keys in the metadata.
-* `to_table()`: returns `nil` or a table with keys:
-    * `fields`: key-value storage
-    * `inventory`: `{list1 = {}, ...}}` (NodeMetaRef only)
-* `from_table(nil or {})`
-    * Any non-table value will clear the metadata
-    * See [Node Metadata] for an example
-    * returns `true` on success
+* `to_table()`:
+    * Returns metadata as a metadata table (see below), if possible
+    * Return `nil` on failure
+* `from_table(data)`
+    * Imports metadata from a metadata table
+    * If `data` is a metadata table (see below),
+      the metadata it represents will replace all metadata
+      of this MetaDataRef object
+    * Any non-table value for `data` will clear all metadata
+    * Returns `true` on success
 * `equals(other)`
     * returns `true` if this metadata has the same key-value pairs as `other`
 
+### Metadata tables
+
+A MetaDataRef can be converted to and from a metadata table,
+which is metadata represented in a table form. It is used
+to serialize and deserialize metadata.
+The methods `from_table` and `to_table` use metadata tables, as
+well as a few callback functions.
+
+A metadata table is a table that has the following keys:
+
+* `fields`: key-value storage of metadata fields
+    * all values are stored as strings
+    * numbers must be converted to strings first
+* `inventory` (for NodeMetaRef only): A node inventory in table form
+    * inventory table keys are inventory list names
+    * inventory table values are item tables
+    * item table keys are slot IDs (starting with 1)
+    * item table values are itemstrings
+    * empty item slots *must* be `""`
```

Comment:
Anything convertible to an ItemStack is fine for `from_table` (i.e. also an ItemStack, a table or `nil`). (Even `nil` should be fine, as `read_items` iterates with `lua_next`. Maybe discourage it though.)

For `to_table`, it should be ItemStacks, as `push_items` calls `LuaItemStack::create`. I haven't tested though.

Reply:
Also, the description could be simpler here, i.e.:
```
* `inventory` is a table that maps listnames to itemlists
* An itemlists is a list of `ItemStack`s, or anything convertible to it. The index (beginning at 1) specifies the inventory list slot.
```